_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+impact of each deployment.
+
+While CI/CD for application code is well understood in the software industry, CI/CD for infrastructure code is a
+relatively new concept. This guide focuses on providing an overview of the background info, design, and implementation
+of a production-ready CI/CD pipeline for infrastructure code.
+
+
+=== What you'll learn in this guide
+
+This guide consists of four main sections:
+
+<<core_concepts>>::
+  An overview of the core concepts you need to understand what a typical CI/CD pipeline entails for appliction and
+  infrastructure code, including a sample workflow, infrastructure to support CI/CD, and threat models to consider to
+  protect your infrastructure.
+
+<<production_grade_design>>::
+  An overview of how to configure a secure, scalable, and robust CI/CD workflow that you can rely on for your
+  production application and infrastructure code. To get a sense of what production-grade means, check out
+  link:/guides/foundations/how-to-use-gruntwork-infrastructure-as-code-library#production_grade_infra_checklist[The production-grade infrastructure checklist].
+
+<<deployment_walkthrough>>::
+  A step-by-step guide to deploying a production-grade CI/CD pipeline in AWS using code from the Gruntwork
+  Infrastructure as Code Library.
+
+<<next_steps>>::
+  What to do once you've got your CI/CD pipeline set up.
+
+
+=== What this guide will not cover
+
+CI/CD for infrastructure code is a large topic and a single guide can not cover everything there is to the topic. As
+such there are several items that this guide will not cover, including:
+
+A pipeline for setting up new environments::
+  This guide will focus on a CI/CD workflow for making changes to infrastructure in an environment that is already set
+  up. In other words, the design and implementation of the pipeline covered in this guide intentionally does not solve
+  the use case of infrastructure code for setting up an environment from scratch. Setting up new environments typically
+  require complex deployment orders and permissions modeling that complicate the task. This makes it hard to automate in
+  a reasonable fashion that still respects the threat model we cover here.
+
+Automated testing and feature toggling strategies for infrastructure code::
+  An important factor of CI/CD pipelines is the existence of automated testing and feature toggles. Automated tests give
+  you the confidence in the code before it is deployed to production. Similarly, feature toggles allow you to partially
+  integrate and deploy code for a feature without enabling it. By doing so, you are able to continuously integrate
+  lareger new developments over time. This guide will briefly introduce automated testing and feature toggles for
+  infrastructure code, but will not do a deep dive on the subject.
+
+
+[[core_concepts]]
+== Core Concepts
+
+[[why_is_it_important_to_have_cicd]]
+=== Why is it important to have CI/CD?
```

Comment:
I love this too! I can't take any credit for this since this was largely taken verbatim from Jim's excellent post https://blog.ycombinator.com/agility-requires-safety/#part1